### ISSUE TITLE
False positive for RUF009 when using imported NewType declarations in dataclass

### ISSUE THREAD

**Issue by MrBhop** [astral-sh/ruff#24202]
### Summary

Hi, I seem to have a false positive for RUF009 when using an imported type, declared with `NewType`. This [comment](https://github.com/astral-sh/ruff/issues/6447#issuecomment-1671887960) from a few years ago suggests, that a fix for this might not be feasible, is that still the case?

pseudo_types.py:
```py
from typing import NewType

MyType = NewType("MyType", float)
```

test_ruff.py:
```py
from dataclasses import dataclass

from pseudo_types import MyType

@dataclass
class MyClass:
    opt: MyType = MyType(0)
```

bug.toml:
```toml
select = [
	"RUF"
]
```

output:
```bash
$ uvx ruff --version
ruff 0.15.7

$ uvx ruff --config=bug.toml check
RUF009 Do not perform function call `MyType` in dataclass defaults
 --> test_ruff.py:7:19
  |
5 | @dataclass
6 | class MyClass:
7 |     opt: MyType = MyType(0)
  |                   ^^^^^^^^^
  |

Found 1 error.
```

### Version

ruff 0.15.7

**Comment by MichaReiser** [comment]
This is a limitation of Ruff. Ruff doesn't perform any analysis across files. That means, Ruff has no understanding of what `MyType` really points to because it doesn't follow the import statement to the `NewType` definition. 

We're working on ty, our type checker, which does support multifile analysis and we have plans to use its more powerful analysis in Ruff in the future, which would allow us to fix this false positive. 

For now, you have to use an inline suppression or you can add `MyType` to immutable calls (https://docs.astral.sh/ruff/settings/#lint_flake8-bugbear_extend-immutable-calls)
